pkg/cloud/services/compute/availabilityzone.go

```diff
@@ -30,3 +30,21 @@ func (s *Service) GetAvailabilityZones() ([]availabilityzones.AvailabilityZone,
 
 	return availabilityZoneList, nil
 }
+
+// GetAvailableAvailabilityZones does as it says, but only returns usable AZs.
+func (s *Service) GetAvailableAvailabilityZones() ([]availabilityzones.AvailabilityZone, error) {
```

Comment:
Couple of things:

Firstly there's only 1 caller of GetAvailabilityZones() and this PR replaces it with a call to this new function, leaving the old one as dead code. We should just modify GetAvailabilityZones() to do what we want.

Secondly, I'm wondering what the impact of removing the FD from cluster.Status.FailureDomains will be in the context of https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1466. In that PR I'm modifying the machine controller to actually look up the failure domain on first use. With this change, if a MachineDeployment referenced an AZ which became unavailable, the cluster controller would eventually delete it. Prior to that the machine would be created but would be unschedulable and enter the ERROR state. After that the machine controller would refuse to create it in the first place and enter a backoff loop waiting for it to exist.

Meh, don't worry about the second one: it's not in scope of your PR. In general, it points out that we have some baked assumptions about AZs being static that aren't necessarily true. We should likely ensure that the cluster controller runs periodically even without k8s watch events. Perhaps every 10 minutes or so.

Reply:
I figured it best to leave both versions, then the caller man make the decision whether they want to consider available ones, or not.  Combining the two views you can derive either {available, not available, non-existent}.  But I'm happy to clean up.

I thought the AZ mutability would possibly raise some eye-brows 😄  I figured, seeing as it's implicitly refreshed every reconcile, then technically speaking it's already mutable, which as you say reacts nicely to AZs vanishing.